HADOOP-16445. Allow separate custom signing algorithms for S3 and DDB

sidseth opened the pull request:
This is very similar to the original patch.
In terms of testing - have run tests against a bucket in us-east-1 (including with the patch posted on HADOOP-16477). Struggling a bit with failures though, which seem completely unrelated to the patch. Still trying to get my test configuration file to a state where tests pass.

hadoop-yetus commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 97 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1454 | trunk passed |
| +1 | compile | 36 | trunk passed |
| +1 | checkstyle | 25 | trunk passed |
| +1 | mvnsite | 37 | trunk passed |
| +1 | shadedclient | 820 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 31 | trunk passed |
| 0 | spotbugs | 71 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 70 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 45 | the patch passed |
| +1 | compile | 36 | the patch passed |
| +1 | javac | 36 | the patch passed |
| +1 | checkstyle | 21 | hadoop-tools/hadoop-aws: The patch generated 0 new + 33 unchanged - 1 fixed = 33 total (was 34) |
| +1 | mvnsite | 36 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 851 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 27 | the patch passed |
| +1 | findbugs | 71 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 86 | hadoop-aws in the patch passed. |
| +1 | asflicense | 32 | The patch does not generate ASF License warnings. |
| | | 3865 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1332 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 0fb3ea60aedb 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 2ae7f44 |
| Default Java | 1.8.0_222 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/1/testReport/ |
| Max. process+thread count | 434 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 90 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1255 | trunk passed |
| +1 | compile | 35 | trunk passed |
| +1 | checkstyle | 25 | trunk passed |
| +1 | mvnsite | 39 | trunk passed |
| +1 | shadedclient | 795 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 25 | trunk passed |
| 0 | spotbugs | 58 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 56 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 32 | the patch passed |
| +1 | compile | 26 | the patch passed |
| +1 | javac | 26 | the patch passed |
| +1 | checkstyle | 20 | hadoop-tools/hadoop-aws: The patch generated 0 new + 33 unchanged - 1 fixed = 33 total (was 34) |
| +1 | mvnsite | 34 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 864 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 25 | the patch passed |
| +1 | findbugs | 71 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 89 | hadoop-aws in the patch passed. |
| +1 | asflicense | 30 | The patch does not generate ASF License warnings. |
| | | 3590 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1332 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 306f19c4b5ff 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:55:24 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / ee7c261 |
| Default Java | 1.8.0_222 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/2/testReport/ |
| Max. process+thread count | 306 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 79 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1193 | trunk passed |
| +1 | compile | 30 | trunk passed |
| +1 | checkstyle | 23 | trunk passed |
| +1 | mvnsite | 35 | trunk passed |
| +1 | shadedclient | 794 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 26 | trunk passed |
| 0 | spotbugs | 58 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 56 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 32 | the patch passed |
| +1 | compile | 26 | the patch passed |
| +1 | javac | 26 | the patch passed |
| +1 | checkstyle | 18 | hadoop-tools/hadoop-aws: The patch generated 0 new + 33 unchanged - 1 fixed = 33 total (was 34) |
| +1 | mvnsite | 30 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 839 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 22 | the patch passed |
| +1 | findbugs | 60 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 67 | hadoop-aws in the patch passed. |
| +1 | asflicense | 29 | The patch does not generate ASF License warnings. |
| | | 3441 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1332 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 83581ae0122f 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:55:24 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / b69ac57 |
| Default Java | 1.8.0_222 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/3/testReport/ |
| Max. process+thread count | 353 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 64 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1175 | trunk passed |
| +1 | compile | 38 | trunk passed |
| +1 | checkstyle | 25 | trunk passed |
| +1 | mvnsite | 40 | trunk passed |
| +1 | shadedclient | 730 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 28 | trunk passed |
| 0 | spotbugs | 60 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 58 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 36 | the patch passed |
| +1 | compile | 28 | the patch passed |
| +1 | javac | 28 | the patch passed |
| +1 | checkstyle | 23 | hadoop-tools/hadoop-aws: The patch generated 0 new + 33 unchanged - 1 fixed = 33 total (was 34) |
| +1 | mvnsite | 32 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 842 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 27 | the patch passed |
| +1 | findbugs | 65 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 80 | hadoop-aws in the patch passed. |
| +1 | asflicense | 31 | The patch does not generate ASF License warnings. |
| | | 3415 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1332 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 29262084f437 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 915cbc9 |
| Default Java | 1.8.0_222 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/4/testReport/ |
| Max. process+thread count | 415 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/4/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 42 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1124 | trunk passed |
| +1 | compile | 34 | trunk passed |
| +1 | checkstyle | 26 | trunk passed |
| +1 | mvnsite | 38 | trunk passed |
| +1 | shadedclient | 720 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 27 | trunk passed |
| 0 | spotbugs | 59 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 57 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 35 | the patch passed |
| +1 | compile | 27 | the patch passed |
| +1 | javac | 27 | the patch passed |
| +1 | checkstyle | 19 | hadoop-tools/hadoop-aws: The patch generated 0 new + 31 unchanged - 1 fixed = 31 total (was 32) |
| +1 | mvnsite | 32 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 749 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 26 | the patch passed |
| +1 | findbugs | 64 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 75 | hadoop-aws in the patch passed. |
| +1 | asflicense | 33 | The patch does not generate ASF License warnings. |
| | | 3227 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1332 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 714a7ba8d01b 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / e04b8a4 |
| Default Java | 1.8.0_222 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/5/testReport/ |
| Max. process+thread count | 447 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/5/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 46 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1116 | trunk passed |
| +1 | compile | 32 | trunk passed |
| +1 | checkstyle | 29 | trunk passed |
| +1 | mvnsite | 42 | trunk passed |
| +1 | shadedclient | 736 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 28 | trunk passed |
| 0 | spotbugs | 62 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 60 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 34 | the patch passed |
| +1 | compile | 32 | the patch passed |
| +1 | javac | 32 | the patch passed |
| +1 | checkstyle | 24 | hadoop-tools/hadoop-aws: The patch generated 0 new + 31 unchanged - 1 fixed = 31 total (was 32) |
| +1 | mvnsite | 35 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 748 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 28 | the patch passed |
| +1 | findbugs | 61 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 80 | hadoop-aws in the patch passed. |
| +1 | asflicense | 31 | The patch does not generate ASF License warnings. |
| | | 3253 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/6/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1332 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 1c1df7add7e1 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / e04b8a4 |
| Default Java | 1.8.0_222 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/6/testReport/ |
| Max. process+thread count | 412 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/6/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

steveloughran commented:
we also use the credentials for talking to STS (session credentials) and I have an ambition to talk to Amazon SQS to subscribe to changes in an S3 bucket for spark streaming. As a result, I'm thinking about how to make that possible after your changes.

I'm also slowly trying to stop `S3AUtils` getting any worse as the one stop "throw all our static static stuff in here.  `S3AUtils` is a merge troublespot and has grown too big. I'm reluctant to move existing stuff, because it makes backporting so hard, but I'd like to make the amount of new stuff we had there nearly non-existent. In particular #970 is adding a new class `org.apache.hadoop.fs.s3a.impl.NetworkBinding` where networking stuff can go; I'm not sure the best place for authentication stuff. Maybe a class alongside that `org.apache.hadoop.fs.s3a.impl.AwsConfigurationFactory`. What do you think?

Also, rather than a new method `createAwsConfForDdb` alongside the existing one, I'd rather than existing one was extended to take a string declaring what the configurations for, e.g : "s3", "ddb", "sts" ... I'm proposing a string over an enum to maintain binary compatibility for applications which add/use a new service. If the name is unknown, we could just warn and return the "default" configuration.

So how about adding the new operations into some into some `AwsConfigurationFactory.createAwsConf(string purpose,...)` method, with `S3AUtils.createAwsConf()` tagged as deprecated and invoking the new method. I'm reluctant to cut it, as I suspect some people (me) have been using it elsewhere.

With that, I'm now going to to some minor review of the patch at the line-by-line level. Those comments come secondary to what I've just suggested.

steveloughran commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java:
```diff
@@ -300,6 +300,8 @@ public void initialize(URI name, Configuration originalConf)
     // clone the configuration into one with propagated bucket options
     Configuration conf = propagateBucketOptions(originalConf, bucket);
+    // Initialize any custom signers
+    initCustomSigners(conf);
```
I guess this is the first thing which needs to be done, before anything else even thinks about talking to AWS services. If so, this is probably the right place. But I would put it after `patchSecurityCredentialProviders`

sidseth replied:
Fixed. Have moved this to just before bindAWSClient

steveloughran commented on hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3AConfiguration.java:
```diff
@@ -19,17 +19,24 @@
 import com.amazonaws.services.s3.AmazonS3;
 import com.amazonaws.services.s3.S3ClientOptions;
 
+import java.io.IOException;
```
should go at the top In its own little block

sidseth replied:
Removed this altogether. Re-factored the tests out into a separate class.

steveloughran replied:
actually, given there's an existing java. block down below, just move it to at least be consistent

steveloughran commented on hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3AConfiguration.java:
```diff
@@ -19,17 +19,24 @@
 import org.apache.hadoop.fs.contract.ContractTestUtils;
 import org.apache.hadoop.fs.s3native.S3xLoginHelper;
 import org.apache.hadoop.test.GenericTestUtils;
+import org.junit.Assert;
```
FWIW, We are moving to assertJ when we think it makes for better assertions.

sidseth replied:
Switched over to AssertJ

steveloughran commented on hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3AConfiguration.java:
```diff
@@ -617,4 +624,135 @@ public void testSecurityCredentialPropagationEndToEnd() throws Exception {
         "override,base");
   }
 
+  @Test(timeout = 10_000L)
```
There's already a test timeout ruleeout so you don't need `timeout = 10_000`.

sidseth replied:
Moved to a separate unit test which doesn't need to extend any other class - so retaining the timeout.

steveloughran commented on hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3AConfiguration.java:
```diff
@@ -617,4 +624,135 @@ public void testSecurityCredentialPropagationEndToEnd() throws Exception {
+    Assert.assertEquals(s3SignerOverride,
+        clientConfiguration.getSignerOverride());
+    clientConfiguration = S3AUtils.createAwsConfForDdb(config, "dontcare");
+    Assert.assertNull(clientConfiguration.getSignerOverride());
```
Needs an error message. FWIW, when writing tests you suspect I will be reviewing, always add an error message. one possible exception, assertEquals -but even there I would prefer one.

When coding, Think to yourself "Jenkins just failed and all I have is this stack trace of the failure -is that enough to debug what went wrong?". If not: log more and expand the assertion message.

sidseth replied:
Done as part of AssertJ

steveloughran commented on hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3AConfiguration.java:
```diff
@@ -617,4 +624,135 @@ public void testSecurityCredentialPropagationEndToEnd() throws Exception {
+  // For now, this is an NPE
+  @Test(timeout = 10_000L, expected = Exception.class)
+  public void testCustomSignerFailureIfNotRegistered() {
+    Signer s1 = SignerFactory.createSigner("testsigner1", null);
```
Prefer `LambaTestUtils.intercept` which lets you assert the text of the caught exception, and which includes the toString value of the result of the l-expression when it did not fail

sidseth replied:
Changed to use LambdaUtils. Not asserting any spcific string though to keep this generic.

steveloughran commented on hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3AConfiguration.java:
```diff
@@ -617,4 +624,135 @@ public void testSecurityCredentialPropagationEndToEnd() throws Exception {
+    initCustomSigners(config);
+    Signer s1 = SignerFactory.createSigner("testsigner1", null);
+    s1.sign(null, null);
+    Assert.assertEquals(true, SignerForTest1.initialized);
```
assertTrue, with error message.

sidseth replied:
Fixed, here and elsewhere.

steveloughran commented on hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3AConfiguration.java:
```diff
@@ -617,4 +624,135 @@ public void testSecurityCredentialPropagationEndToEnd() throws Exception {
+
+    Signer s2 = SignerFactory.createSigner("testsigner2", null);
+    s2.sign(null, null);
+    Assert.assertEquals(true, SignerForTest2.initialized);
```
assertTrue, with error message.

sidseth commented:
Removed the new method in S3AUtil, and introduced a SignerManager (more changes coming to this soon). Also incorporated one of the changes from the patch on HADOOP-16505 which uses the correct way to initialize signers.

On the AwsConfigurationFactory - that makes sense. However, I don't think that should be in this patch. It's unrelated, and will end up moving more code from S3AUtils (including some public static methods). That's better handled in a separate refactoring only patch.

sidseth commented:
Updated. Have left the timeouts on the test, since these tests don't need the default 10 minute timeout.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 39 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1024 | trunk passed |
| +1 | compile | 35 | trunk passed |
| +1 | checkstyle | 28 | trunk passed |
| +1 | mvnsite | 39 | trunk passed |
| +1 | shadedclient | 739 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 30 | trunk passed |
| 0 | spotbugs | 59 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 57 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 33 | the patch passed |
| +1 | compile | 27 | the patch passed |
| -1 | javac | 27 | hadoop-tools_hadoop-aws generated 2 new + 15 unchanged - 0 fixed = 17 total (was 15) |
| -0 | checkstyle | 20 | hadoop-tools/hadoop-aws: The patch generated 16 new + 34 unchanged - 1 fixed = 50 total (was 35) |
| +1 | mvnsite | 32 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 791 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 29 | the patch passed |
| +1 | findbugs | 76 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 96 | hadoop-aws in the patch passed. |
| +1 | asflicense | 33 | The patch does not generate ASF License warnings. |
| | | 3240 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/7/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1332 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux ca9b87211251 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 01d7924 |
| Default Java | 1.8.0_222 |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/7/artifact/out/diff-compile-javac-hadoop-tools_hadoop-aws.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/7/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/7/testReport/ |
| Max. process+thread count | 415 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/7/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

steveloughran commented on hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestSignerManager.java:
```diff
@@ -0,0 +1,124 @@
+
+  @Rule
+  public Timeout testTimeout = new Timeout(
+10_000L, TimeUnit.MILLISECONDS
```
indentation.
FWIW, for any ITest, the values in S3ATestConstants are the ones to refer to, not any other ones. Why so? When you are debugging things in the IDE you can edit those values in one place and know that your debug session won't suddenly finish 5-10 minutes in. I usually end up doing exactly that after my first debug session times out

sidseth replied:
That is set to 10 minutes, which is a little too high for simple tests like this. A test which actually gets stuck would end up wasting a lot of time and resources when running on Jenkins.

steveloughran commented on hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestSignerManager.java:
```diff
@@ -0,0 +1,124 @@
+import com.amazonaws.auth.Signer;
+import com.amazonaws.auth.SignerFactory;
+import java.util.concurrent.TimeUnit;
+import org.apache.hadoop.classification.InterfaceAudience.Private;
```
all org.aoache stuff needs to go in its own block after the block of non-asf imports

sidseth replied:
Fixed in the next patch. I couldn't find anything on the coding standards, or a standard IntelliJ template.

steveloughran commented on hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestSignerManager.java:
```diff
@@ -0,0 +1,124 @@
+
+package org.apache.hadoop.fs.s3a;
+
+import static org.apache.hadoop.fs.s3a.Constants.CUSTOM_SIGNERS;
```
all statics in the same block at the bottom of the imports

sidseth replied:
Changed.

sidseth commented:
Had forgotten how difficult it is to get a patch into Hadoop. Are there any published coding standards or templates for IntelliJ / Eclipse that you're aware of?

There's multiple revisions of the patch just fixing checkstyles, which is rather annoying.

steveloughran commented:
> Had forgotten how difficult it is to get a patch into Hadoop. Are there any published coding standards or templates for IntelliJ / Eclipse that you're aware of?

not AFAIK. Spark has some automated checks for import ordering but as there's no IDE config which works 100% of the time there either, also a PITA. But as it is 100% consistent, it is probably better at avoiding a lot of backport-merge-hell, which imports always seem to generate

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 162 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1293 | trunk passed |
| +1 | compile | 34 | trunk passed |
| +1 | checkstyle | 29 | trunk passed |
| +1 | mvnsite | 40 | trunk passed |
| +1 | shadedclient | 809 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 31 | trunk passed |
| 0 | spotbugs | 73 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 69 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 38 | the patch passed |
| +1 | compile | 32 | the patch passed |
| -1 | javac | 32 | hadoop-tools_hadoop-aws generated 2 new + 15 unchanged - 0 fixed = 17 total (was 15) |
| +1 | checkstyle | 21 | hadoop-tools/hadoop-aws: The patch generated 0 new + 34 unchanged - 1 fixed = 34 total (was 35) |
| +1 | mvnsite | 38 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 869 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 27 | the patch passed |
| +1 | findbugs | 68 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 86 | hadoop-aws in the patch passed. |
| +1 | asflicense | 32 | The patch does not generate ASF License warnings. |
| | | 3791 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.2 Server=19.03.2 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/8/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1332 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux b6fdcb367432 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / e97f0f1 |
| Default Java | 1.8.0_222 |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/8/artifact/out/diff-compile-javac-hadoop-tools_hadoop-aws.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/8/testReport/ |
| Max. process+thread count | 411 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/8/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 100 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 4 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1208 | trunk passed |
| +1 | compile | 32 | trunk passed |
| +1 | checkstyle | 23 | trunk passed |
| +1 | mvnsite | 37 | trunk passed |
| +1 | shadedclient | 806 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 24 | trunk passed |
| 0 | spotbugs | 59 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 57 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 34 | the patch passed |
| +1 | compile | 29 | the patch passed |
| +1 | javac | 29 | the patch passed |
| +1 | checkstyle | 21 | hadoop-tools/hadoop-aws: The patch generated 0 new + 42 unchanged - 1 fixed = 42 total (was 43) |
| +1 | mvnsite | 35 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 875 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 23 | the patch passed |
| +1 | findbugs | 64 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 85 | hadoop-aws in the patch passed. |
| +1 | asflicense | 31 | The patch does not generate ASF License warnings. |
| | | 3582 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.2 Server=19.03.2 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/9/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1332 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 712a7a1f798a 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / f16eb09 |
| Default Java | 1.8.0_212 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/9/testReport/ |
| Max. process+thread count | 310 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/9/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

sidseth commented:
Tested against us-east-2 and ap-south-1 (-Ds3guard -Ddynamo -Dauth). There's a standard set of test failures with and without the patch.
```
[ERROR]   ITestMagicCommitMRJob>AbstractITCommitMRJob.testMRJob:146->AbstractFSContractTestBase.assertIsDirectory:327 » FileNotFound
[ERROR]   ITestDirectoryCommitMRJob>AbstractITCommitMRJob.testMRJob:146->AbstractFSContractTestBase.assertIsDirectory:327 » FileNotFound
[ERROR]   ITestPartitionCommitMRJob>AbstractITCommitMRJob.testMRJob:146->AbstractFSContractTestBase.assertIsDirectory:327 » FileNotFound
[ERROR]   ITestStagingCommitMRJob>AbstractITCommitMRJob.testMRJob:146->AbstractFSContractTestBase.assertIsDirectory:327 » FileNotFound
[ERROR]   ITestDynamoDBMetadataStore.setUp:159 » TestTimedOut test timed out after 10000...
[ERROR]   ITestS3GuardConcurrentOps.testConcurrentTableCreations:178->deleteTable:87 » IllegalArgument
```

steveloughran commented:
last two are intermittent and we should have a go at fixing (longer timeouts, more retries). 

other ones are new. I will have to run locally to see if they occur for me, and if not, assume its some config quirk.
Do bear in mind those MR test jobs are spawning new processes. Anything you do for FS test setup in the JUnit process has no effect there

steveloughran commented:
tests all went well for me

steveloughran commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/SignerManager.java:
```diff
@@ -0,0 +1,99 @@
+import com.amazonaws.auth.SignerFactory;
+import java.io.Closeable;
+import java.io.IOException;
+import org.apache.hadoop.conf.Configuration;
```
imports still a bit confused here

sidseth replied:
Ouch. Hopefully fixed in the latest commit.

steveloughran commented:
I think you've managed to get stuff into the latest commit which weren't there earlier (SignerManager). Was that your intent? Because that's a bigger change. I'm happy with the smaller patch and its details; lets get that in before the next step

sidseth commented:
The change the latest commit (SignerManager) removes the code from S3AUtils and moves it to a SignerManager class (non-static). The actual signer registration has picked up bits from the patch on HADOOP-16505 (duplicate of this jira), since that was doing the registration in a better way (once per unique signer per JVM, rather than each time a FS instances is created). 

I do plan to make additional changes to this class in a follow up patch, but it's essentially the same as the old patch, with a non-static Closeable class, without adding an extra static method to S3AUtils. Please let me know if there's specific concerns around this (i.e. change back to what?)

steveloughran commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/auth/STSClientFactory.java:
```diff
@@ -31,6 +31,7 @@
 import com.amazonaws.services.securitytoken.model.Credentials;
 import com.amazonaws.services.securitytoken.model.GetSessionTokenRequest;
 import com.google.common.base.Preconditions;
+import org.apache.hadoop.fs.s3a.Constants;
```
import location

hadoop-yetus commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 98 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 4 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1192 | trunk passed |
| +1 | compile | 32 | trunk passed |
| +1 | checkstyle | 24 | trunk passed |
| +1 | mvnsite | 36 | trunk passed |
| +1 | shadedclient | 793 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 25 | trunk passed |
| 0 | spotbugs | 56 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 55 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 34 | the patch passed |
| +1 | compile | 27 | the patch passed |
| +1 | javac | 27 | the patch passed |
| +1 | checkstyle | 19 | hadoop-tools/hadoop-aws: The patch generated 0 new + 42 unchanged - 1 fixed = 42 total (was 43) |
| +1 | mvnsite | 31 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 825 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 22 | the patch passed |
| +1 | findbugs | 60 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 81 | hadoop-aws in the patch passed. |
| +1 | asflicense | 29 | The patch does not generate ASF License warnings. |
| | | 3485 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=18.09.7 Server=18.09.7 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/10/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1332 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 15cf078856e3 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / f260b5a |
| Default Java | 1.8.0_222 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/10/testReport/ |
| Max. process+thread count | 308 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/10/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

sidseth commented:
> tests all went well for me
Do you have fs.s3a.metadatastore.authoritative, fs.s3a.metadatastore.impl set in core-site.xml or auth-keys.xml

The MR tests seem quite broken to me otherwise. They use cluster configuration, which is created from new JobConf() - this will not pick up the test properties. Can get at least 2 of them to pass (haven't tried the other 2) with these 2 config keys set in core-site.xml or auth-keys.xml.
That said, I'm seeing references to "part-m-00000", "part-m-00001", "_SUCCESS" in S3Guard - with a 0 size, and marked as deleted, and cannot explain where these come from. Maybe some of the committer code runs on the client.

steveloughran commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/Constants.java:
```diff
@@ -343,9 +343,43 @@ private Constants() {
+   */
+  public static final String CUSTOM_SIGNERS = "fs.s3a.custom.signers";
+
+  /**
```
you are going to have to add these to the aws docs I'm afraid. This is probably time to start a new "signing" section

steveloughran commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/auth/delegation/SessionTokenBinding.java:
```diff
@@ -31,6 +31,7 @@
 import com.amazonaws.auth.AWSSessionCredentials;
 import com.amazonaws.services.securitytoken.AWSSecurityTokenService;
 import com.google.common.annotations.VisibleForTesting;
+import org.apache.hadoop.fs.s3a.Constants;
```
import location

steveloughran commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/DynamoDBClientFactory.java:
```diff
@@ -27,6 +27,7 @@
 import com.amazonaws.services.dynamodbv2.AmazonDynamoDBClientBuilder;
 import com.google.common.base.Preconditions;
 import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.fs.s3a.Constants;
```
import location

steveloughran commented on hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestSignerManager.java:
```diff
@@ -0,0 +1,129 @@
+import com.amazonaws.auth.AWSCredentials;
+import com.amazonaws.auth.Signer;
+import com.amazonaws.auth.SignerFactory;
+import java.util.concurrent.TimeUnit;
```
go on, pull this up into its own block...

steveloughran commented on hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestSignerManager.java:
```diff
@@ -0,0 +1,129 @@
+/**
+ * Tests for the SignerManager.
+ */
+public class TestSignerManager {
```
Going to highlight that the completely optional org.apache.hadoop.test.HadoopTestBase class can act as a base class, offering setup of the test timeouts and automatic naming of the junit thread to the current method. I would encourage your use of it.

steveloughran commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/SignerManager.java:
```diff
@@ -0,0 +1,99 @@
+
+import com.amazonaws.auth.Signer;
+import com.amazonaws.auth.SignerFactory;
+import java.io.Closeable;
```
usual and predictable comments about imports

steveloughran commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/Constants.java:
```diff
@@ -343,9 +343,43 @@ private Constants() {
+   * algorithm. fs.s3a.signing-algorithm - This property has existed for the
+   * longest time. If specified, without either of the other 2 properties being
+   * specified, this signing algorithm will be used for S3 and DDB (S3Guard).
+   * The other 2 properties override this value for S3 or DDB.
```
also: other uses like STS.

steveloughran commented:
Patch LGTM, only thing needed is to get the imports consistent. I know its a PITA but its to keep merge hell vaguely under control.

You are going to need to document the new signing options in the markdown, rather than just the source code -either here or in a followup JIRA. That would also be good time to list what the standard AWS signers are as I don't see any real docs of them at all.

sidseth commented:
Updated the imports again. They should be consistent now.
```
java.*
\n
org.apache.*
\n
Everything else
\n
import static *
```
is the formatting options I'm using.

I'll update the docs in the next jira, which will make some changes to the value set in some of the config options.

Does the LGTM qualify as a +1? I'll go ahead and squash-merge the patch if that's the case.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 39 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 4 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1041 | trunk passed |
| +1 | compile | 36 | trunk passed |
| +1 | checkstyle | 29 | trunk passed |
| +1 | mvnsite | 40 | trunk passed |
| -1 | shadedclient | 110 | branch has errors when building and testing our client artifacts. |
| +1 | javadoc | 31 | trunk passed |
| 0 | spotbugs | 60 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 57 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 34 | the patch passed |
| +1 | compile | 28 | the patch passed |
| +1 | javac | 28 | the patch passed |
| +1 | checkstyle | 21 | hadoop-tools/hadoop-aws: The patch generated 0 new + 42 unchanged - 1 fixed = 42 total (was 43) |
| +1 | mvnsite | 32 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| -1 | shadedclient | 29 | patch has errors when building and testing our client artifacts. |
| +1 | javadoc | 24 | the patch passed |
| +1 | findbugs | 59 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 82 | hadoop-aws in the patch passed. |
| +1 | asflicense | 26 | The patch does not generate ASF License warnings. |
| | | 1815 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/11/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1332 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux e64d223753e4 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / b3173e1 |
| Default Java | 1.8.0_222 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/11/testReport/ |
| Max. process+thread count | 342 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1332/11/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

sidseth commented:
Also, the *MRJob tests pass after applying HADOOP-16591 or setting configs in auth-keys.xml

sidseth commented:
/retest

steveloughran commented:
+1

sidseth commented:
Thanks for the reviews @steveloughran . Merging.
The shadedclient failure isn't related. See: https://github.com/apache/hadoop/pull/1432#issuecomment-531755823
